### PR TITLE
✅ Pull-to-Refresh implementation

### LINKED ISSUE #50: 💡 User's schedule data is not updated whenever data from i-Maluum is updated
# Description
Whenever new data from i-Maluum emerges, user need to remove all the relative subjects and re-add it again to see the changes.

# Steps to Reproduce
1. Create a new schedule by adding subjects or courses.
2. (At this particular time, there are some unassigned lecturers / lecture rooms for some courses).
3. Changes from i-Maluum didn't reflect on the current version of the app.
4. Users need to remove all subjects and re-add it again to see new data emerges.

# More Information
Device: OnePlus 9 (LE2115)
OS: OxygenOS 12 (Android 12)
App Version: 1.0.3

Reported by @haikalakif

# Attachments

### PR DESCRIPTION
**Pull request by ElyasAsmad**
# Description

Refreshes the schedule data from _schedule1.php_

Fixes #50 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Update Preview
p-t-r iiumschedule

### CODE REVIEW

File: lib/views/saved_schedule/saved_schedule_layout.dart [hunk 1]
```diff
@@ -42,6 +61,82 @@ class _SavedScheduleLayoutState extends State<SavedScheduleLayout> {
...
+       */
+      int colour = 0xFF000000 | (Random().nextInt(0xFFFFFF));
+      return SavedSubject.fromSubject(
+        subject: subject,
+        subjectName: subject.title,
+        hexColor: colour
+      );
+    }).toList();
+    currentSchedule.lastModified = DateTime.now().toString();
+    widget._box.put(scheduleIdx, currentSchedule);
```
Comment (iqfareez) [inline lib/views/saved_schedule/saved_schedule_layout.dart:136]: Can be replaced with 
```
currentSchedule.save();
```

**Review by iqfareez — COMMENTED** [review]
Salam. Thank you for this very nice addition!! 😄 I always wanted to add this feature but couldn't find time to do it.

Suggestion: Skip update subject where venue is empty. For example, in my cases, my lab venue in imaalum is blank. The lecturer update abt the venue through ms teams etc, so I updated manually in the schedule. With this update, seems like it can override my previously set schedule.

File: lib/views/saved_schedule/saved_schedule_layout.dart [hunk 2]
```diff
@@ -42,6 +61,82 @@ class _SavedScheduleLayoutState extends State<SavedScheduleLayout> {
...
+       * 
+       * TODO: Implement colour generator class (maybe use Material 3? Refer #30)
+       */
+      int colour = 0xFF000000 | (Random().nextInt(0xFFFFFF));
+      return SavedSubject.fromSubject(
+        subject: subject,
+        subjectName: subject.title,
+        hexColor: colour
+      );
+    }).toList();
```
[the remark this answers is not in the document]
Comment (ElyasAsmad) [reply lib/views/saved_schedule/saved_schedule_layout.dart:134]: Hello Iqmal,

Regarding the random colour thing, actually I was having some problem while updating the Hive object, as the new fetched data overrides the previous colour used, causing the app to crash as the colour value is null (refer https://github.com/iqfareez/iium_schedule/issues/50#issuecomment-1293314349).

Anyways, I will try to find a workaround for this matter. 😎👍🏻

File: lib/views/saved_schedule/saved_schedule_layout.dart [hunk 3]
```diff
@@ -42,6 +61,82 @@ class _SavedScheduleLayoutState extends State<SavedScheduleLayout> {
...
+    
+    /*
+      NOTE:
+      For now, we only use data from the title field of the SavedSchedule object.
+
+      TODO: Save student's kuliyyah into HiveDB
+    */
+
+    // Get kuliyyah code (e.g: KICT) from tokenized title
+    final kuliyyah = name.split(' ').elementAt(0);
```
[the remark this answers is not in the document]
Comment (ElyasAsmad) [reply lib/views/saved_schedule/saved_schedule_layout.dart:80]: I will try to implement the feature to save kuliyyah into HiveDB. New commit will be submitted afterwards.

File: lib/views/saved_schedule/saved_schedule_layout.dart [hunk 4]
```diff
@@ -42,6 +62,94 @@ class _SavedScheduleLayoutState extends State<SavedScheduleLayout> {
...
+      
+      /**
+       * In some cases, a course venue can be empty in i-Maluum and lecturers will update through
+       * other platforms (i.e: Whatsapp / Microsoft Teams). Since users will override the
+       * venue all by themselves, it is necessary to skip empty venue from i-Maluum
+       * as it will override the current venue.
+       * 
+       * Refer https://github.com/iqfareez/iium_schedule/pull/51#pullrequestreview-1158961053
+       */
+      if (updatedSubject.venue == '') return prevSubject;
```
Comment (iqfareez) [inline lib/views/saved_schedule/saved_schedule_layout.dart:133]:
```suggestion
      if (updatedSubject.venue == null) return prevSubject;
```
Checking if the string is empty doesn't guarantee the venue is empty. I think checking for null is better, in-line with the albiruni implementation:

https://github.com/iqfareez/albiruni/blob/b8c561eb33ca71625b231bcb3fab993ed64a8952/lib/src/albiruni_base.dart#L152

**Review by iqfareez — CHANGES_REQUESTED** [review]

File: lib/views/saved_schedule/saved_schedule_layout.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (iqfareez) [reply lib/views/saved_schedule/saved_schedule_layout.dart]: Nice works. Seems like the random colour thingy has been resolved., right?

**Comment by ElyasAsmad** [comment]
Hello @iqfareez !

In these commits, I have added support for HiveDB to save student's kuliyyah and regenerated Hive's adapter files for this project.

For the #51 feature, I have updated the `onRefresh` method to use the data from HiveDB instead of using the hardcoded kuliyyah information in the schedule title (which is very unsafe).

File: lib/views/saved_schedule/saved_schedule_layout.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (ElyasAsmad) [reply lib/views/saved_schedule/saved_schedule_layout.dart]: Yep! 

[099032d](https://github.com/iqfareez/iium_schedule/pull/51/commits/099032d1813252fb7a0a78176c3d15c2c8e2b518)

File: lib/views/saved_schedule/saved_schedule_layout.dart (hunk 4 above)
Comment (ElyasAsmad) [reply lib/views/saved_schedule/saved_schedule_layout.dart]: Fixed the code, thanks for the new knowledge!

Comment (iqfareez) [reply lib/views/saved_schedule/saved_schedule_layout.dart]: No problem

**Review by iqfareez — APPROVED** [review]
lgtm!

You've Got LGTM

However, before merging this PR, I'm gonna work with #47 first.

**Comment by iqfareez** [comment]
@all-contributors add @ElyasAsmad  for code

**Comment by iqfareez** [comment]
Hrp jadi. @all-contributors add @ElyasAsmad for code

**Pull request merged**